runtime/src/main/java/org/corfudb/runtime/LiteRoutingQueueListener.java

```diff
@@ -72,105 +77,95 @@ public LiteRoutingQueueListener(CorfuStore corfuStore, String sourceSiteId, Stri
                 txnContext.commit();
             }
             log.debug("Deleted {} messages from {}", allEntries.size(), recvQ.getFullyQualifiedTableName());
         }
     }
+
     @Override
     public CorfuStoreMetadata.Timestamp performFullSync() {
-        CorfuStoreMetadata.Timestamp ts = null;
+        CorfuStoreMetadata.Timestamp ts;
         List<Table.CorfuQueueRecord> allMsgs;
+
         try (TxnContext txnContext = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
             allMsgs = txnContext.entryList(recvQ);
             ts = txnContext.commit();
         }
 
         List<Queue.RoutingTableEntryMsg> batchOneOfAKind = new ArrayList<>();
-        // TODO: Fix the enum once the receiver sets the type for full sync as well.
-        Queue.ReplicationType currentType = Queue.ReplicationType.LAST_FULL_SYNC_ENTRY;
         int entriesProcessed = 0;
         for (Table.CorfuQueueRecord q : allMsgs) {
             Queue.RoutingTableEntryMsg msg = (Queue.RoutingTableEntryMsg) q.getEntry();
-            if (!msg.getReplicationType().equals(currentType)) {
-                if (currentType.equals(Queue.ReplicationType.LAST_FULL_SYNC_ENTRY)) {
-                    currentType = msg.getReplicationType();
-                    batchOneOfAKind.add(msg);
-                    continue;
-                }
-                if (currentType.equals(Queue.ReplicationType.LOG_ENTRY_SYNC)) {
-                    log.info("LiteRecvQ::performFullSync {} LOG_ENTRY messages", batchOneOfAKind.size());
-                    if (processUpdatesInLogEntrySync(batchOneOfAKind)) {
-                        entriesProcessed = entriesProcessed + batchOneOfAKind.size();
-                    }
-                } else {
-                    log.info("LiteRecvQ::performFullSync {} SNAPSHOT messages", batchOneOfAKind.size());
-                    if (processUpdatesInSnapshotSync(batchOneOfAKind)) {
-                        entriesProcessed = entriesProcessed + batchOneOfAKind.size();
-                    }
-                } // else we have encountered a change in type, reset our batch..
+
+            if (msg.getReplicationType() != currentReplicationType) {
+                entriesProcessed += processBatch(currentReplicationType, batchOneOfAKind);
+
+                // We have encountered a change in type, reset our batch..
                 batchOneOfAKind = new ArrayList<>();
-                currentType = msg.getReplicationType();
-            } // else batch up message of the same kind so we can deliver in a batch
+                currentReplicationType = msg.getReplicationType();
+
+                // If the current message is of type LAST_SNAPSHOT_SYNC_ENTRY, invoke the completion callback
+                if (msg.getReplicationType() == LAST_SNAPSHOT_SYNC_ENTRY) {
+                    onSnapshotSyncComplete();
+                }
+            }
+            // Batch up message of the same kind so we can deliver in a batch
             batchOneOfAKind.add(msg);
```

Comment:
this should be before processBatch. Maybe on L147?

Reply:
Also, we want to deliver all the snapshot msgs together (as the comment says), then we should have the logic where on seeing a different type of msg is when we call `processBatch`.
Currently processBatch would be called per txn.

Reply:
the check on L148 is when a message of another replication type was received.  So we hand over all messages collected till there.  

The current message is added to the batch after processing in case of replication type mismatch